merlin-server/sql/merlin/plan_collaboration.sql

```diff
@@ -0,0 +1,188 @@
+-- TODO list:
+--   - plan snapshot table (done)
+--   - duplicate function (done)
+--        - duplicate temporal subset of plan
+--   - merge request table
+--   - diff function
+--   - history tracking (tables and function) (done)
+--   - merge base function
+
+create table plan_snapshot_parent(
+  snapshot_id integer,
+  parent_snapshot_id integer
+);
+
+create table plan_latest_snapshot(
+  plan_id integer,
+  snapshot_id integer
+);
+
+-- Snapshot is a collection of the state of all the activities as they were at the time of the snapshot
+-- as well as any other properties of the plan that can change
+create table plan_snapshot(
+  snapshot_id integer generated always as identity,
+
+  plan_id integer not null,
```

Comment:
please add key constraints

Reply:
Will add a primary key notice to snapshot_id, but because snapshots persist after plans are deleted, I believe that excludes plan_id from being a foreign key on plan.

Reply:
Why do snapshots persist?

Reply:
regardless, you can still use `on delete set null` and allow `plan_id` to be null. It wouldn't make sense to reference a deleted plan anyway, right?

Reply:
TL;DR: To not break merges. If I'm currently merging your plan into mine and you delete your plan, I should still be able to finish my merge. Also, the same snapshot can be in the history of several plans (they have a many to many relationship), so it doesn't make sense to delete on plan delete.

Also:
In general, we decided that once a merge request has been made, no changes made to the plan supplying changes will impact what the plan getting merged into sees when they go to merge. Not having snapshots would violate that decision.

Similarly, snapshot history and finding the merge base can break if we have snapshots deleted with their plan. For a simple example, if we:

1. create a plan
2. create a child
3. create a grandchild from the child
4. create a merge request to merge the grandchild into the grandparent plan
5. delete the parent plan
6. attempt to begin the merge between the grandchild and grandparent

Then our merge base, which was the creation snapshot of the child plan, has been unexpectedly deleted. This means that deleting one plan has affected a merge request involving two different plans. Similarly, if we had actually begun the merge between the grandchild and grandparent and the child plan was deleted at some point between when the grandparent plan is locked and `begin_merge`/the UI finish wanting to get data from the merge base snapshot, then we're risking an erroneous state.

Reply:
Ok, that all makes sense. You can still add a foreign key constraint though, because it doesn't make to reference sense a plan that doesn't exist.